offer-manager-domain/src/main/groovy/life/qbic/business/offers/content/CreateOfferContent.groovy

```diff
@@ -0,0 +1,208 @@
+        String creationDate = dateFormat.format(offer.modificationDate)
+        String expirationDate = offer.expirationDate.toLocalDate().toString()
+        groupProductItems(offer.items)
+
+        def offerContent = new OfferContent.Builder(offer.customer,offer.selectedCustomerAffiliation,offer.projectManager,creationDate,expirationDate,offer.projectTitle
+        ,offer.projectObjective, offer.experimentalDesign.orElse(""),offer.identifier.toString())
+
+        //collect productitems and convert to offeritems
+        groupProductItems(offer.items)
+        List<OfferItem> dataManagementOfferItems = dataManagementItems.collect{createProductItemToOfferItem(it)}
+        List<OfferItem> dataAnalysisOfferItems = dataAnalysisItems.collect{createProductItemToOfferItem(it)}
+        List<OfferItem> dataGenerationOfferItems = dataGenerationItems.collect{createProductItemToOfferItem(it)}
+
+        offerContent.dataGenerationItems(dataGenerationOfferItems)
+        .dataAnalysisItems(dataAnalysisOfferItems)
+        .dataManagementItems(dataManagementOfferItems)
+
+        double overheadsDA = calculateOverheadSum(dataAnalysisOfferItems)
+        double overheadsDG = calculateOverheadSum(dataGenerationOfferItems)
+        double overheadsPMandDS = calculateOverheadSum(dataManagementOfferItems)
+
+        offerContent.overheadsDataAnalysis(overheadsDA)
+        .overheadsDataGeneration(overheadsDG)
+        .overheadsProjectManagementAndDataStorage(overheadsPMandDS)
+        .overheadTotal(offer.overheadSum)
+
+        offerContent.netDataAnalysis(calculateNetSum(dataAnalysisOfferItems))
+        .netDataGeneration(calculateNetSum(dataGenerationOfferItems))
+        .netProjectManagementAndDataStorage(calculateNetSum(dataManagementOfferItems))
+        .netCost(offer.totalNetPrice)
+
+        offerContent.totalVat(offer.taxCosts)
+        .vatRatio(offer.determineTaxCost())
+        .totalCost(offer.totalCosts)
+
+        OfferContent offerContentFinal = offerContent.build()
+
+        output.createdOfferContent(offerContentFinal)
+    }
+
+    /**
+     * Helper method that calculates the NET price for a list of product items
+     * @param offerItems The product item list for which the NET is calculated
+     *  @return The net value for the given list of items
+     */
+    private static double calculateNetSum(List<OfferItem> offerItems) {
+        double netSum = 0
+        offerItems.each {
+            netSum += it.quantity * it.unitPrice
+        }
+        return netSum
+    }
+
+    /**
+     * Calculates the overhead sum of all product items
+     * @param offerItems Items for which the overheads are calculated
+     * @return The overhead price for all items
+     */
+    private double calculateOverheadSum(List<OfferItem> offerItems) {
+        double overheadSum = 0.0
+        offerItems.each {
+            overheadSum += it.quantity * it.unitPrice * overheadRatio
```

Comment:
maybe we could extract java.util.BiFunction here for the different calculations.
e.g.
```
class CalculateItemOverhead extends Function<BigDecimal, BigDecimal> {
  private final BigDecimal overheadRatio

  public CalculateItemOverhead(double overheadRatio) {
    this.overheadRatio = new BigDecimal(overheadRatio)
  }

 BigDecimal apply(BigDecimal costs) {
    return costs * overheadRatio
  }

}
```

```
class CalculateItemNet extends BiFunction<BigDecimal, BigDecimal, BigDecimal> {

 BigDecimal apply(BigDecimal quantity, BigDecimal unitPrice) {
    return quantity * unitPrice
  }

}
```

```
class DetermineUnitPrice extends BiFunction<double, double, BigDecimal> {

  private final AffiliationCategory affiliationCategory

  public DetermineUnitPrice(AffiliationCategory affiliationCategory) {
    this.affiliationCategory = affiliationCategory
  }

 BigDecimal apply(BigDecimal internalPrice, BigDecimal externalPrice) {
    if (this.affiliationCategory == AffiliationCategory.INTERNAL) {
      return new BigDecimal(internalPrice)
    } else {
      return new BigDecimal(externalPrice)
    }
  }

}
```

Reply:
then you could compose them in the offer business object and here and be sure that the computation logic is the same

Reply:
@sven1103 what do you think?

Reply:
I am not sure if I get your point. IMO its sufficient as it is. What is the benefit of the bifunction? I think it gives it more complexity as needed

Reply:
the benefit is that both, the offer object and the use case use the same computation and we do not have 2 places where the net sum and the overheads are computed